Visualizing workflow runs with an invocation graph view

ahmedhamidawan opened the pull request:
This adds a **graph view** to the workflow invocation summary, which utilizes the workflow editor canvas and original workflow structure, and instead of showing step details on each step node, it shows the job states for the steps. Clicking on a step expands the details for the invocation step.

The `useInvocationGraph` composable loads the graph onto the readonly editor based on the original workflow id and invocation object; and the step info is loaded via the `step_jobs_summary` api route.

| An example invocation overview |
| ---------------- |
| ![image](https://github.com/galaxyproject/galaxy/assets/78516064/88d4b408-4308-42ff-83e7-8ce60b758f6d) |

## Sample Workflow Run:

https://github.com/galaxyproject/galaxy/assets/78516064/32dc87c5-7cfc-47c5-8fe4-5c042cdc4a83

# Invocation View Header:
This also adds a header with details about the invocation:
![image](https://github.com/galaxyproject/galaxy/assets/78516064/fc2fed46-aaf8-46c2-b452-75d923dad414)

### Older (alternative?) Implementation:
With a Card containing everything inline:
![image](https://github.com/galaxyproject/galaxy/assets/78516064/3ec283db-0983-4713-81a1-790412998589)

# Hide and Show steps as needed:
To view multiple individual steps at the same time (without the graph getting in the way), users can hide/show steps like so:

https://github.com/galaxyproject/galaxy/assets/78516064/74604985-5645-4384-972d-97516baa68f9

# Invocations List Change: Route to invocation instead of only expanding table rows
This change allows you to click an invocation workflow name and route to the dedicated full page (center-panel) summary for the invocation, while you can still expand it within the table by clicking on the expand details icon in the first column:

## Before:

https://github.com/galaxyproject/galaxy/assets/78516064/a9d9d968-d86d-4a0a-b112-a4df8f493c75

## Now:

https://github.com/galaxyproject/galaxy/assets/78516064/8f9585de-999e-4115-8832-89f8d98ba544

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
Add an invocation graph view to the invocation summary

This graph view utilizes the same workflow editor canvas, and instead of showing step details on each step node, it shows the job states for the steps. Clicking on a step expands the details for the invocation step.

The `useInvocationGraph` composable loads the graph based on the original workflow id and invocation object; and the step info is loaded via the `step_jobs_summary` api route.

ahmedhamidawan committed:
add `implicit_collection_jobs_id` to invocation step summary

Also, return a correctly encoded database id field in `step_jobs_summary` api.

ahmedhamidawan committed:
make graph and steps appear fixed side by side, improve job display

ahmedhamidawan committed:
handle skipped/conditional steps in invocation graph view

ahmedhamidawan committed:
improve styling of `View Report` & `Generate PDF` buttons

ahmedhamidawan committed:
route to invocation view instead of expanding within invocation list

This allows users to directly go back and forth between dedicated full-center-panel invocation summaries and the invocations list.

ahmedhamidawan committed:
change `skipped` state color to a lighter shade

ahmedhamidawan committed:
refactor `WorkflowInvocationState` and improve its jest

Added a mock implementation for the store to detect fetches and use those to check if invocation or job states are terminal

ahmedhamidawan committed:
fix jest fails caused by modifications made for invocation graph

ahmedhamidawan committed:
fix selenium by only hiding `label` in `NodeInput` with `props.blank`

Otherwise, seleniums would fail for:
```
Timeout waiting on CSS selector [[node-label='...'] [input-name='...'] + .delete-terminal-button] to become clickable.
```

ahmedhamidawan committed:
render parts of `InvocationGraph` only if `Summary` tab is active

Otherwise, this selenium fails:
`FAILED lib/galaxy_test/selenium/test_workflow_invocation_details.py::TestWorkflowInvocationDetails::test_job_details - selenium.common.exceptions.TimeoutException: Message: Timeout waiting on CSS selector [[data-step="1"] .step-title] to become visible.`
because even if you change tab from Summary to Details, the content is still in the template and the steps in the Summary conflicts with the steps in the Details tab; and are hence not detected in the selenium.

ahmedhamidawan committed:
do not default-expand outputs for `dataStep` in graph view

ahmedhamidawan committed:
do not show tooltip labels for output nodes in invocation graph

ahmedhamidawan committed:
remove hover bg-success for output nodes in invocation graph

ahmedhamidawan committed:
move steps to `WorkflowInvocationSteps` in a card; improve styling

Scroll to activeId when clicked (on mounted, also expand first step that has a failed (`error`) state

ahmedhamidawan committed:
fix `activeNodeId` not expanding step if it is = 0

Also removes the redundant prop as the value can be local to `WorkflowInvocationSteps`

ahmedhamidawan committed:
load full workflow for the version that was run

Without this, the latest version (with potential missing steps from the version that was run) is used to create the graph, and therefore, due to potential mismatch between versions, the graph is incorrect.

ahmedhamidawan committed:
add an invocation view button to individual `InvocationsList`

...i.e.: the `InvocationsList` for an individual workflow, which doesn't have the workflow name.

ahmedhamidawan committed:
center tooltip for `SwitchToHistoryLink`

ahmedhamidawan committed:
add fixed header to `WorkflowInvocationState`, scroll job to view

ahmedhamidawan committed:
add `isFullPage` prop instead of checking route

ahmedhamidawan committed:
rename Summary to Overview, remove Steps tab, add tabs for params

ahmedhamidawan committed:
show invocation count in invocation state view as well

ahmedhamidawan committed:
rename `showStep` to `showJob`

ahmedhamidawan committed:
always toggle a non-input step in graph view

If there isn't an errored step, still toggle/expand the last step in the workflow invocation.

ahmedhamidawan committed:
make it obvious that jobs are clickable and viewable at the bottom

By always showing the card at the bottom, with instructions on how to view a job.

ahmedhamidawan committed:
only expand last step if the invocation is terminal

ahmedhamidawan committed:
add tooltip to `ShowGraph` button

ahmedhamidawan committed:
add icon for queued state in invocation graph

ahmedhamidawan committed:
change card to div for invocation view header and steps

ahmedhamidawan committed:
if invocation tab is scrollable, add a `pr-2` for scrollbar

This is to make the multiple scrollbars in `WorkflowInvocationOverview` to be clickable (one for the Steps and one for the whole tab)

davelopez commented on client/src/components/Workflow/Invocation/Graph/InvocationGraph.vue:
```diff
@@ -0,0 +1,327 @@
+
+interface Props {
+    /** The invocation to display */
+    invocation: components["schemas"]["WorkflowInvocationElementView"];
```
One minor general refactoring recommendation about importing API models in the frontend:

Instead of using components["schemas"], try to import those types or interfaces from `@/api`. In particular, for this model, you can find it in `@/api/invocations`, or you can export new ones if they are not already defined. This approach offers several advantages IMHO, such as better readability and the ability to update the model name in one place if it changes in the backend for any reason. Also, within those modules, we define some additional interfaces that may be useful, either by combining several interfaces or by defining missing interfaces while we migrate more schema models in the backend.

ahmedhamidawan replied:
I have used (and added to) the api schemas instead. Thank you for spotting this!

ahmedhamidawan committed:
change history on invocation success without reloading

Also, change the `href` for multiview in this view to a <router-link> (to prevent reload).

ahmedhamidawan committed:
clear shown job regardless of `props.fullPage`

Also, do not show a separate input steps section if there is only one input step.

ahmedhamidawan committed:
use `FlexPanel` for steps in `InvocationGraph`

ahmedhamidawan committed:
fix bugs with `stateStore.activeNodeId` not syncing

ahmedhamidawan committed:
fix API model imports, remove unused function from `useInvocationGraph`

ahmedhamidawan committed:
try to fix scroll-to by using `scrollTo` instead of `scrollIntoView`

ahmedhamidawan committed:
fix selenium so that it collapses graph to expand step, fix height of steps if graph is collapsed, by removing the max-height

ElectronicBlueberry commented on client/src/components/Workflow/Editor/Node.vue:
```diff
@@ -219,6 +241,23 @@ const style = computed(() => {
+        }
+    }
+    return cls;
+});
```
defining computed classes as a `Record<string, number>` has the advantage of vastly simplifying code like this.

This change also requires changes to:
- useInvocationGraph.ts line 198
- useInvocationGraph.ts line 38

```suggestion
const headerClass = computed(() => {
    return {
        ...invocationStep.value.headerClass,
        "cursor-pointer": props.isInvocation,
        "node-header": !props.isInvocation || invocationStep.value.headerClass === undefined,
        "cursor-move": !props.readonly && !props.isInvocation,
    };
});
```

ahmedhamidawan replied:
Yes, that is better, though don't you mean `Record<string, boolean>` instead of `Record<string, number>`?

ahmedhamidawan committed:
Update client/src/components/Workflow/Editor/Node.vue

Co-authored-by: Laila Los <44241786+ElectronicBlueberry@users.noreply.github.com>

ahmedhamidawan committed:
change `headerClass` to an object instead of a string

Co-authored-by: Laila Los <44241786+ElectronicBlueberry@users.noreply.github.com>

dannon committed:
Minor fixes to Heading component; correct imports and default props.

dannon committed:
Merge remote-tracking branch 'upstream/dev' into graph_view_invocations

mvdbeek reviewed (approved):
This is so cool, I've been wanting something like this ever since I ran my first workflow on Galaxy!
